File: src/main/java/com/github/stars/service/SimilarRepoService.java
package com.github.stars.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.stars.entity.GithubRepo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import javax.annotation.Resource;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * 相似项目发现服务
 * 基于 GitHub Search API，根据当前仓库特征搜索相似的高质量项目
 */
@Service
public class SimilarRepoService {

    private static final Logger log = LoggerFactory.getLogger(SimilarRepoService.class);
    private static final String GITHUB_API = "https://api.github.com";
    private static final int MAX_RESULTS = 20;
    private static final int MIN_STARS = 100;

    @Resource
    private GithubRepoService githubRepoService;

    @Resource
    private SystemConfigService configService;

    @Resource
    private ObjectMapper objectMapper;

    @Autowired
    @Qualifier("longTimeoutRestTemplate")
    private RestTemplate restTemplate;

    /**
     * 搜索结果项
     */
    public static class SimilarRepo {
        public String fullName;
        public String description;
        public String language;
        public int stars;
        public int forks;
        public String htmlUrl;
        public String pushedAt;
        public String aiReason; // AI 推荐理由
        public double score;    // 综合评分
    }

    /**
     * 发现与指定仓库相似的优质项目
     */
    public List<SimilarRepo> findSimilar(Long repoId) {
        GithubRepo source = githubRepoService.findById(repoId);
        if (source == null) return Collections.emptyList();

        List<SimilarRepo> allResults = new ArrayList<>();

        // 策略1: 按 topic 搜索
        List<String> topics = parseTopics(source.getTopics());
        for (String topic : topics) {
            if (allResults.size() >= MAX_RESULTS) break;
            List<SimilarRepo> results = searchGitHub("topic:" + topic, source.getFullName());
            allResults.addAll(results);
        }

        // 策略2: 按语言+关键词搜索
        if (allResults.size() < MAX_RESULTS && source.getLanguage() != null) {
            String query = "language:" + source.getLanguage();
            if (source.getRepoName() != null) {
                // 提取仓库名中的关键词（去掉常见后缀）
                String name = source.getRepoName().replaceAll("[-_.].*", "");
                if (name.length() >= 3) query += " " + name;
            }
            List<SimilarRepo> results = searchGitHub(query, source.getFullName());
            allResults.addAll(results);
        }

        // 去重 + 排除当前仓库
        Set<String> seen = new HashSet<>();
        List<SimilarRepo> unique = new ArrayList<>();
        for (SimilarRepo r : allResults) {
            if (r.fullName.equals(source.getFullName())) continue;
            if (seen.add(r.fullName)) unique.add(r);
        }

        // 按评分排序
        unique.sort((a, b) -> Double.compare(b.score, a.score));

        // 限制返回数量
        if (unique.size() > MAX_RESULTS) {
            unique = unique.subList(0, MAX_RESULTS);
        }

        // 用 AI 生成推荐理由
        if (!unique.isEmpty()) {
            enrichWithAI(unique, source);
        }

        log.info("为 {} 发现 {} 个相似项目", source.getFullName(), unique.size());
        return unique;
    }

    /**
     * 调用 GitHub Search API
     */
    private List<SimilarRepo> searchGitHub(String query, String excludeRepo) {
        List<SimilarRepo> results = new ArrayList<>();
        try {
            // 构建完整查询：排除自身 + 3个月活跃 + ≥100 stars
            String threeMonthsAgo = LocalDate.now().minusMonths(3).format(DateTimeFormatter.ISO_DATE);
            String fullQuery = query + " stars:>=" + MIN_STARS + " pushed:>" + threeMonthsAgo;

            String url = UriComponentsBuilder.fromHttpUrl(GITHUB_API + "/search/repositories")
                    .queryParam("q", fullQuery)
                    .queryParam("sort", "stars")
                    .queryParam("order", "desc")
                    .queryParam("per_page", 10)
                    .build()
                    .toUriString();

            HttpHeaders headers = new HttpHeaders();
            headers.set("Accept", "application/vnd.github.v3+json");
            headers.set("User-Agent", "GithubStars-SimilarFinder");
            String ghToken = configService.getValue("github.token");
            if (ghToken != null && !ghToken.isEmpty()) {
                headers.set("Authorization", "Bearer " + ghToken);
            }

            HttpEntity<String> entity = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                JsonNode root = objectMapper.readTree(response.getBody());
                JsonNode items = root.get("items");
                if (items != null && items.isArray()) {
                    for (JsonNode item : items) {
                        SimilarRepo r = new SimilarRepo();
                        r.fullName = item.get("full_name").asText();
                        r.description = item.has("description") && !item.get("description").isNull()
                                ? item.get("description").asText() : "";
                        r.language = item.has("language") && !item.get("language").isNull()
                                ? item.get("language").asText() : "";
                        r.stars = item.get("stargazers_count").asInt();
                        r.forks = item.get("forks_count").asInt();
                        r.htmlUrl = item.get("html_url").asText();
                        r.pushedAt = item.has("pushed_at") && !item.get("pushed_at").isNull()
                                ? item.get("pushed_at").asText() : "";
                        // 综合评分：Star为主要因子
                        r.score = Math.log10(r.stars + 1) * 10 + Math.log10(r.forks + 1) * 2;
                        results.add(r);
                    }
                }
            } else if (response.getStatusCode() == HttpStatus.FORBIDDEN) {
                log.warn("GitHub Search API 限流");
            }
        } catch (Exception e) {
            log.warn("GitHub Search 失败: {} - {}", query, e.getMessage());
        }
        return results;
    }

    /**
     * 使用 AI 为每个相似项目生成一句推荐理由
     */
    private void enrichWithAI(List<SimilarRepo> repos, GithubRepo source) {
        // 只对前 10 个生成 AI 推荐理由（节省 token）
        List<SimilarRepo> toEnrich = repos.size() > 10 ? repos.subList(0, 10) : repos;
        if (toEnrich.isEmpty()) return;

        try {
            StringBuilder sb = new StringBuilder();
            sb.append("源项目: ").append(source.getFullName()).append("\n");
            sb.append("描述: ").append(source.getDescriptionCn() != null ? source.getDescriptionCn() : source.getDescription()).append("\n\n");
            sb.append("以下是搜索到的相似项目，请为每个项目用一句话说明它为什么值得关注（15字以内）:\n\n");
            for (int i = 0; i < toEnrich.size(); i++) {
                SimilarRepo r = toEnrich.get(i);
                sb.append(i + 1).append(". ").append(r.fullName)
                        .append(" (Star:").append(r.stars)
                        .append(", Lang:").append(r.language).append(")\n");
                if (r.description != null && !r.description.isEmpty()) {
                    sb.append("   描述:").append(r.description.length() > 100
                            ? r.description.substring(0, 100) + "..." : r.description).append("\n");
                }
            }
            sb.append("\n返回格式: JSON数组 [\"项目1的推荐理由\", \"项目2的推荐理由\", ...]，只返回JSON数组，不要其他内容。");

            // 调用 DeepSeek
            JsonNode response = callDeepSeek(sb.toString());
            if (response != null) {
                JsonNode choices = response.get("choices");
                if (choices != null && choices.isArray() && choices.size() > 0) {
                    String content = choices.get(0).get("message").get("content").asText().trim();
                    // 清理 markdown 代码块
                    if (content.startsWith("```")) {
                        content = content.replaceAll("```json\\s*", "").replaceAll("```\\s*", "");
                    }
                    try {
                        List<String> reasons = objectMapper.readValue(content, new TypeReference<List<String>>() {});
                        for (int i = 0; i < Math.min(reasons.size(), toEnrich.size()); i++) {
                            toEnrich.get(i).aiReason = reasons.get(i);
                        }
                    } catch (Exception e) {
                        log.warn("AI推荐理由解析失败: {}", e.getMessage());
                    }
                }
            }
        } catch (Exception e) {
            log.warn("AI 推荐理由生成失败: {}", e.getMessage());
        }
    }

    private JsonNode callDeepSeek(String prompt) {
        try {
            com.fasterxml.jackson.databind.node.ObjectNode requestBody = objectMapper.createObjectNode();
            requestBody.put("model", configService.getValue("deepseek.model", "deepseek-chat"));
            requestBody.put("temperature", 0.5);
            requestBody.put("max_tokens", 2048);

            com.fasterxml.jackson.databind.node.ArrayNode messages = objectMapper.createArrayNode();
            com.fasterxml.jackson.databind.node.ObjectNode sysMsg = objectMapper.createObjectNode();
            sysMsg.put("role", "system");
            sysMsg.put("content", "你是GitHub项目推荐专家。只返回要求格式的JSON，不说任何废话。");
            messages.add(sysMsg);

            com.fasterxml.jackson.databind.node.ObjectNode userMsg = objectMapper.createObjectNode();
            userMsg.put("role", "user");
            userMsg.put("content", prompt);
            messages.add(userMsg);
            requestBody.set("messages", messages);

            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            headers.set("Authorization", "Bearer " + configService.getValue("deepseek.api_key"));

            HttpEntity<String> entity = new HttpEntity<>(objectMapper.writeValueAsString(requestBody), headers);
            ResponseEntity<String> resp = restTemplate.postForEntity(
                    configService.getValue("deepseek.api_url", "https://api.deepseek.com/v1/chat/completions"),
                    entity, String.class);

            if (resp.getStatusCode() == HttpStatus.OK && resp.getBody() != null) {
                return objectMapper.readTree(resp.getBody());
            }
        } catch (Exception e) {
            log.warn("DeepSeek 调用失败: {}", e.getMessage());
        }
        return null;
    }

    private List<String> parseTopics(String topicsJson) {
        if (topicsJson == null || topicsJson.isEmpty() || "[]".equals(topicsJson)) {
            return Collections.emptyList();
        }
        try {
            return objectMapper.readValue(topicsJson, new TypeReference<List<String>>() {});
        } catch (Exception e) {
            return Collections.emptyList();
        }
    }
}

File: backend-nestjs/src/ai/services/similar-repo.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '../../config/config.service';
import { GithubRepoService } from '../../github/services/github-repo.service';
import { GithubApiService } from '../../github/services/github-api.service';

const MAX_RESULTS = 20;
const MIN_STARS = 100;

@Injectable()
export class SimilarRepoService {
    private readonly logger = new Logger(SimilarRepoService.name);
    constructor(
        private readonly config: ConfigService,
        private readonly githubRepo: GithubRepoService,
        private readonly githubApi: GithubApiService,
    ) {}

    /**
     * 调用 DeepSeek API（P1-11 FIX: 加入 system message 角色提示）
     *
     * @param prompt 用户提示词
     * @param systemMsg 可选的系统角色消息
     * @returns AI 返回的文本内容，失败返回 null
     */
    private async callDeepSeek(prompt: string, systemMsg?: string): Promise<string | null> {
        const apiKey = await this.config.getValue('deepseek.api_key');
        const apiUrl = await this.config.getValueDefault('deepseek.api_url', 'https://api.deepseek.com/v1/chat/completions');
        const model = await this.config.getValueDefault('deepseek.model', 'deepseek-chat');
        if (!apiKey) return null;
        const msgs: any[] = [];
        if (systemMsg) msgs.push({ role: 'system', content: systemMsg });
        msgs.push({ role: 'user', content: prompt });
        try {
            const res = await fetch(apiUrl, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${apiKey}` },
                body: JSON.stringify({ model, temperature: 0.5, max_tokens: 2048, messages: msgs }),
            });
            if (!res.ok) return null;
            const data = (await res.json()) as any;
            return data.choices?.[0]?.message?.content?.trim() || null;
        } catch {
            return null;
        }
    }

    /**
     * 解析仓库的 topics 字段（JSON 字符串 -> 字符串数组）
     *
     * @param topicsStr topics JSON 字符串
     * @returns topics 字符串数组，解析失败返回空数组
     */
    private parseTopics(topicsStr: string | null): string[] {
        if (!topicsStr) return [];
        try {
            return JSON.parse(topicsStr);
        } catch {
            return [];
        }
    }

    /**
     * 通过 AI 从仓库描述和 README 中提取搜索关键词（P1-11 FIX: 加入 system message）
     *
     * 请求 DeepSeek 提取 3~5 个英文搜索关键词，用于 GitHub Search API 检索相似仓库。
     * 若 AI 提取失败，回退为从仓库名中提取关键词。
     *
     * @param repo 仓库信息对象
     * @returns 搜索关键词数组
     */
    private async extractKeywords(repo: any): Promise<string[]> {
        const desc = String(repo.descriptionCn || repo.description || '');
        const readme = String(repo.readmeCn || repo.readmeOriginal || '').substring(0, 600);
        const prompt = `提取 3~5 个英文搜索关键词。只返回 JSON 数组:\n\n项目名: ${repo.fullName}\n描述: ${desc}\nREADME: ${readme || '无'}\n\n输出: ["keyword1", "keyword2"]`;
        try {
            const result = await this.callDeepSeek(prompt, '你是GitHub项目推荐专家。只返回要求格式的JSON，不说任何废话。');
            if (result) {
                const cleaned = result
                    .replace(/^```json\s*/i, '')
                    .replace(/^```\s*/i, '')
                    .replace(/```$/i, '')
                    .trim();
                const keywords = JSON.parse(cleaned) as string[];
                if (Array.isArray(keywords) && keywords.length > 0) return keywords;
            }
        } catch {
            this.logger.warn('AI关键词提取失败');
        }
        return [(repo.repoName || repo.fullName || '').replace(/[-_.].*/, '')];
    }

    /**
     * 通过 GitHub Search API 搜索仓库并转换为统一格式（P1-12 FIX: pushed:>= 而非 pushed:>）
     *
     * 搜索最近 3 个月内推送过且超过 {MIN_STARS} star 的仓库。
     * 同时计算一个综合评分（score）用于排序。
     *
     * @param query 搜索关键词
     * @returns 格式化后的仓库列表
     */
    private async searchAndConvert(query: string): Promise<any[]> {
        const d = new Date();
        d.setMonth(d.getMonth() - 3);
        const dateStr = d.toISOString().split('T')[0];
        const fullQuery = `${query} stars:>=${MIN_STARS} pushed:>=${dateStr}`;
        const items = await this.githubApi.searchRepos(fullQuery, 'stars', 'desc', 10);
        return items.map((item: any) => ({
            fullName: item.full_name || '',
            description: item.description || '',
            language: item.language || '',
            stars: item.stargazers_count || 0,
            forks: item.forks_count || 0,
            htmlUrl: item.html_url || '',
            pushedAt: item.pushed_at || '',
            aiReason: '',
            score: Math.log10((item.stargazers_count || 0) + 1) * 10 + Math.log10((item.forks_count || 0) + 1) * 2,
        }));
    }

    /**
     * 为搜索结果生成 AI 推荐理由（P1-12 FIX: 加入 system message）
     *
     * 取前 10 个结果，请求 DeepSeek 为每个仓库生成一句推荐理由。
     * 理由写入对应仓库对象的 aiReason 字段。
     *
     * @param repos 搜索结果列表
     * @param source 源仓库信息
     */
    private async enrichWithAI(repos: any[], source: any) {
        if (!repos.length) return;
        const top = repos.slice(0, 10);
        let list = top.map((r, i) => `${i}. ${r.fullName} ⭐${r.stars} - ${r.language} - ${r.description.substring(0, 100)}`).join('\n');
        const prompt = `源项目: ${source.fullName} (${source.descriptionCn || source.description || ''})\n\n相似项目:\n${list}\n\n为每个项目写一句推荐理由，返回 JSON 数组: ["理由1", "理由2"]`;
        try {
            const result = await this.callDeepSeek(prompt, '你是GitHub项目推荐专家。只返回要求格式的JSON，不说任何废话。');
            if (result) {
                const reasons = JSON.parse(
                    result
                        .replace(/^```json\s*/i, '')
                        .replace(/```$/i, '')
                        .trim(),
                ) as string[];
                if (Array.isArray(reasons))
                    top.forEach((r, i) => {
                        r.aiReason = reasons[i] || '';
                    });
            }
        } catch {
            this.logger.warn('AI推荐理由生成失败');
        }
    }

    /**
     * 查找与指定仓库相似的项目
     *
     * 流程：提取关键词 → 通过 GitHub Search API 搜索 → 补充 topic 搜索 → 排序评分 → AI 生成推荐理由。
     * 每次搜索最多返回 {MAX_RESULTS} 个结果，最多 {MIN_STARS} star 以上。
     *
     * @param repoId 源仓库 ID
     * @returns 相似仓库列表（含评分和 AI 推荐理由）
     */
    async findSimilar(repoId: number): Promise<any[]> {
        this.logger.log('查找相似仓库: repoId=' + repoId);
        const source = await this.githubRepo.findById(repoId);
        if (!source) {
            this.logger.warn('源仓库未找到: repoId=' + repoId);
            return [];
        }
        const seen = new Set<string>();
        seen.add(source.fullName || '');
        const keywords = await this.extractKeywords(source);
        let allResults: any[] = [];

        for (const kw of keywords) {
            if (allResults.length >= MAX_RESULTS * 2) break;
            const q = kw + (source.language ? ` language:${source.language}` : '');
            for (const r of await this.searchAndConvert(q)) {
                if (!seen.has(r.fullName)) {
                    seen.add(r.fullName);
                    allResults.push(r);
                }
            }
        }

        if (allResults.length < MAX_RESULTS) {
            const topics = this.parseTopics(source.topics);
            for (const topic of topics.slice(0, 3)) {
                if (allResults.length >= MAX_RESULTS) break;
                for (const r of await this.searchAndConvert(`topic:${topic}`)) {
                    if (!seen.has(r.fullName)) {
                        seen.add(r.fullName);
                        allResults.push(r);
                    }
                }
            }
        }

        allResults.sort((a, b) => b.score - a.score);
        const topResults = allResults.slice(0, MAX_RESULTS);
        if (topResults.length > 0) await this.enrichWithAI(topResults, source);
        this.logger.log('相似仓库查找完成: repoId=' + repoId + ', resultCount=' + topResults.length);
        return topResults;
    }
}
